### PR TITLE
[BOOTSECT][FREELDR] Support booting from NTFS partitions

### PR DESCRIPTION
**Pull request by sdever**
## Purpose

Implement NTFS boot sector that loads FreeLdr from a NTFS partition

Jira issue: [CORE-17474](https://jira.reactos.org/browse/CORE-17474)

## Proposed changes

 - Implement VBR for NTFS
 - Fix a bug in FreeLdr's NTFS driver preventing it to find freeldr.inf
 
## TODO

- Handle "weird" NTFS partition with uncommon sector sizes / sectors per cluster / sectors per index record
- Better implementation for file loading
- 64-bit LCN support
- More testing

### CODE REVIEW

**Comment by JoachimHenze** [comment]
Great! How can we test that?

**Comment by sdever** [comment]
> Great! How can we test that?

I manually copied the ntfs.bin into the partition (with the command `dd if=boot/freeldr/bootsect/ntfs.bin of=/dev/loop0p1 bs=1 skip=84 seek=84` on Linux), and copied the ReactOS bootcd and freeldr.sys on the mounted NTFS partition

I didn't change usetup yet to install the VBR on NTFS partition, as someone else said he worked on NTFS formatting/install on Mattermost.

File: boot/freeldr/bootsect/ntfs.S
Comment (HBelusca) [inline boot/freeldr/bootsect/ntfs.S:389]: Is the extra level of indentation on purpose?

File: boot/freeldr/bootsect/ntfs.S
Comment (HBelusca) [inline boot/freeldr/bootsect/ntfs.S:304]: Please add an information block along the lines of what is done in the FAT32 bootsector:
https://git.reactos.org/?p=reactos.git;a=blob;f=boot/freeldr/bootsect/fat32.S;h=ff9c40af1211a7f7112b0a74d2bc7cdd876a8a0d;hb=HEAD#l317

Also, please check whether on Windows the NTFS boot sector is stored on disk contiguously or not. If not, please also add a comment similar to the one written in the FAT32 bootsector, concerning the second part that is (in FAT32 case) at sector 14.

File: boot/freeldr/bootsect/ntfs.S
Comment (HBelusca) [inline boot/freeldr/bootsect/ntfs.S:801]: Maybe add and use a "Loading FreeLdr..." message as we do with our other boot sectors, see example here:
https://git.reactos.org/?p=reactos.git;a=blob;f=boot/freeldr/bootsect/fat32.S;h=ff9c40af1211a7f7112b0a74d2bc7cdd876a8a0d;hb=HEAD#l530

File: boot/freeldr/bootsect/ntfs.S
Comment (HBelusca) [inline boot/freeldr/bootsect/ntfs.S:48]: Unused by NTFS ?

**Review by HBelusca — COMMENTED** [review]
I know that Windows' NTFS bootsector code complains if the boot loader is on a compressed NTFS volume. Do you intend to implement such check (and complain in this situation)?

File: boot/freeldr/bootsect/ntfs.S
Comment (sdever) [reply boot/freeldr/bootsect/ntfs.S:48]: Those values are used by NTFS, but can't be computed at compile time (since it depends of partition scheme). The labels are here only to get correct addresses to access those values in the code.

**Comment by sdever** [comment]
> I know that Windows' NTFS bootsector code complains if the boot loader is on a compressed NTFS volume. Do you intend to implement such check (and complain in this situation)?

This is on my todo-list yes (I forgot to add it in PR todo list)

File: boot/freeldr/bootsect/ntfs.S
Comment (sdever) [reply boot/freeldr/bootsect/ntfs.S:304]: Boot sector is stored on disk contiguously on my Windows partitions: it's represented by the $Boot file (inode number 7), and uses the two first clusters in the first partition as marked in the run lists. Also, linux-ntfs documentation states this file cannot be relocated: https://flatcap.org/linux-ntfs/ntfs/files/boot.html.

I'll add what I wrote here as comment for the boot sector

File: boot/freeldr/bootsect/ntfs.S
Comment (HBelusca) [reply boot/freeldr/bootsect/ntfs.S:48]: My comment was only about "TotalSectorsBig"; according to documentation (e.g. Starman's MBR/VBR stuff, and another website) the TotalSectorsBig doesn't seem to be used by NTFS.

Comment (sdever) [reply boot/freeldr/bootsect/ntfs.S:48]: Yes, TotalSectorsBig is set to 0 on NTFS partitions formatted by Windows (and I was misled by Github display of PRs :p)

File: boot/freeldr/bootsect/ntfs.S
Comment (Extravert-ir) [inline boot/freeldr/bootsect/ntfs.S:59]: There is `.quad` macro for 64-bit numbers

File: boot/freeldr/bootsect/ntfs.S
Comment (mzji) [inline boot/freeldr/bootsect/ntfs.S:826]: `freeldr.sys is a sparse, compressed or encrypted file, cannot boot` ?

File: boot/freeldr/bootsect/ntfs.S
Comment (mzji) [inline boot/freeldr/bootsect/ntfs.S:309]: IIRC cluster size can vary between different NTFS instances? Thus two clusters don't really means 8 sectors in all cases, right? Or only NTFS instances those have 1 cluster = 4 sectors can be used for boot? BTW 1 sector on PC HDD usually is 512 bytes in the old days, and a NTFS cluster by default is 4KiB, that means a NTFS cluster by default equals 8 sectors, not 4 sectors.

Comment (sdever) [reply boot/freeldr/bootsect/ntfs.S:309]: For the moment, it only boots on sectors with 512 bytes, as it's the most common NTFS settings, see the TODO list in the pull request.
Also I messed up with my calculations, it's 16 sectors for two clusters, thanks for pointing it :)

File: boot/freeldr/bootsect/ntfs.S
Comment (Extravert-ir) [inline boot/freeldr/bootsect/ntfs.S:326]: I would put a comment here, this for example https://flatcap.org/linux-ntfs/ntfs/files/boot.html

> This can be negative, which means that the size of the MFT/Index record is smaller than a cluster. In this case the size of the MFT/Index record in bytes is equal to 2^(-1 * Clusters per MFT/Index record). So for example if Clusters per MFT Record is 0xF6 (-10 in decimal), the MFT record size is 2^(-1 * -10) = 2^10 = 1024 bytes.

Just a nit, this feels more straightforward:
```
test al, al
js NegativeOffset
```

File: boot/freeldr/bootsect/ntfs.S
Comment (Extravert-ir) [inline boot/freeldr/bootsect/ntfs.S:334]: what is `8` here?

File: boot/freeldr/bootsect/ntfs.S
Comment (Extravert-ir) [inline boot/freeldr/bootsect/ntfs.S:438]: The `near` can be omitted?

File: boot/freeldr/bootsect/ntfs.S
Comment (Extravert-ir) [inline boot/freeldr/bootsect/ntfs.S:451]: Hex and Dec numbers are mixed

File: boot/freeldr/bootsect/ntfs.S
Comment (Extravert-ir) [inline boot/freeldr/bootsect/ntfs.S:536]: Case-insensitive compare

File: boot/freeldr/bootsect/ntfs.S
Comment (Extravert-ir) [inline boot/freeldr/bootsect/ntfs.S:574]: Maybe just store a multiplied number too?

File: boot/freeldr/bootsect/ntfs.S
Comment (Extravert-ir) [inline boot/freeldr/bootsect/ntfs.S:584]: Just make it `eax`, that way you can avoid `movzx eax, ax` stuff (it is already called like that actually)

File: boot/freeldr/bootsect/ntfs.S
Comment (Extravert-ir) [inline boot/freeldr/bootsect/ntfs.S:596]:
```suggestion
    add   eax, dword ptr [MFTStartSector]  // Add it to the start of the MFT
```

Comment (Extravert-ir) [reply boot/freeldr/bootsect/ntfs.S:596]: And `cx` will not be touched

File: boot/freeldr/bootsect/ntfs.S
Comment (Extravert-ir) [inline boot/freeldr/bootsect/ntfs.S:602]: I don't know NTFS that well. What is this?

File: boot/freeldr/bootsect/ntfs.S
Comment (Extravert-ir) [inline boot/freeldr/bootsect/ntfs.S:750]: This is a return value as well

File: boot/freeldr/bootsect/ntfs.S
Comment (Extravert-ir) [inline boot/freeldr/bootsect/ntfs.S:652]: This looks a bit unusual. Why not `DS:AX` and `ES:BX` (to fit the `ReadCluster`)?

File: boot/freeldr/bootsect/ntfs.S
Comment (Extravert-ir) [inline boot/freeldr/bootsect/ntfs.S:660]: Wasting bytes :)

**Review by Extravert-ir — DISMISSED** [review]
Good job!
I didn't dive that much into NTFS-specific stuff here.
Please, answer the comments and format functions like I did in the BTRFS VBR:
```
// Searches a key in a BTRFS tree
// INPUT:
// - [bp+4] BTRFS key to find
// - EDX:EAX logical address of root header
// - CL leaf node level
// OUTPUT:
// - ES:[SI] pointer to found item's key
// - ES:[BX] pointer to found item
// - ECX item length
// LOCALS:
// - [bp-8] - block number/current node offset
// - [bp-9] - current level
SearchTree:
```

File: boot/freeldr/bootsect/ntfs.S
Comment (HBelusca) [reply boot/freeldr/bootsect/ntfs.S:602]: It's `ELIF`  (aka. `FILE` written in little endian).

File: boot/freeldr/bootsect/ntfs.S
Comment (HBelusca) [inline boot/freeldr/bootsect/ntfs.S:19]: (Regarding a previous comment by @Extravert-ir ): Yes the publicly available documentation on the format of the NTFS boot sector shows that all these fields are indeed part of the NTFS BPB.

File: boot/freeldr/bootsect/ntfs.S
Comment (sdever) [reply boot/freeldr/bootsect/ntfs.S:320]: I should learn how to perform 64-bit calculations in real-mode assembly before  (which is quite a PITA compared to protected mode), so I'll handle 64 bit in a future iteration

File: boot/freeldr/bootsect/ntfs.S
Comment (sdever) [reply boot/freeldr/bootsect/ntfs.S:438]: MSVC build whines if not present

File: boot/freeldr/bootsect/ntfs.S
Comment (sdever) [reply boot/freeldr/bootsect/ntfs.S:652]: Because I didn't want to touch DS, since it was useful to print debug messages on that part

File: boot/freeldr/bootsect/ntfs.S
Comment (sdever) [reply boot/freeldr/bootsect/ntfs.S:602]: I should make a define here yes

**Comment by Zero3K** [comment]
I tried it on my 2006 Macbook. I get a ? Folder icon instead of it booting to ReactOS.

**Comment by Zero3K** [comment]
I have an issue with booting NTFS on a 2006 Macbook after applying this and the other two patches needed to get ReactOS to use MS' NTFS driver. I wrote the bootsector via dd under Slax and copied over the ReactOS directory and boot files from the USB that I copied them to (from a VHD that I installed the modified ReactOS to. When booting via the SSD after doing that, Loading FreeLdr... keeps being spammed.

File: boot/freeldr/bootsect/ntfs.S
Comment (HBelusca) [reply boot/freeldr/bootsect/ntfs.S:439]: This would then be our first asm code with near ptr.
Could you see whether by removing `near ptr` it works then? (that's how all of our other code does it; alternatively it uses `short`)

File: boot/freeldr/bootsect/ntfs.S
Comment (aaSSfxxx) [reply boot/freeldr/bootsect/ntfs.S:608]: We can't, since we don't know what cluster we'll need to read

File: boot/freeldr/bootsect/ntfs.S
Comment (sdever) [reply boot/freeldr/bootsect/ntfs.S]: For some reason MSVC stopped whining on rhis and compiles fine without.

File: boot/freeldr/bootsect/ntfs.S
Comment (sdever) [reply boot/freeldr/bootsect/ntfs.S]: ReadCluster takes an arbitrary cluster number, and it needs to be converted to sectors then, so it can't be precomputed.

File: boot/freeldr/bootsect/ntfs.S
Comment (HBelusca) [reply boot/freeldr/bootsect/ntfs.S]: Maybe it's part of their new set of bugs they "introduced" in MASM when it compiles 16-bit code?
See another instance of this there: https://developercommunity.visualstudio.com/t/MASMML-x86:-Inconsistent-large-mov/1385890

File: boot/freeldr/bootsect/ntfs.S
Comment (sdever) [reply boot/freeldr/bootsect/ntfs.S:257]: This is weird, without the offset, the emitted code dereferences the symbol instead of taking its address...

Comment (HBelusca) [reply boot/freeldr/bootsect/ntfs.S:257]: What's the MASM/ML version you are using? If it's one from VS 2019, could you try testing with an older VS version if possible? Maybe it's yet another one of their new "bugs"... 🤔

**Comment by HBelusca** [comment]
**Rebased** PR on master after PR #3775 being committed.

**Comment by binarymaster** [comment]
@JoachimHenze reopening would only make sense if you're going to handle the PR yourself (and self-assign it).

It's not assigned to anyone so far, so it will be closed once again in 4 months.

**Comment by JoachimHenze** [comment]
We have one devs approval so far. That might encourage a seocnd one. At least that is my hope. Otherwise I will accordingly to our guidelines push the SquashnMerge button in a week from now.
The only critics in here so far is missing documentation (extravert). That could be added retrospectively as well. And imho doesn't justify blocking the PR.

**Review by JoachimHenze — APPROVED** [review]

File: boot/freeldr/freeldr/lib/fs/ntfs.c
Comment (HBelusca) [inline boot/freeldr/freeldr/lib/fs/ntfs.c:736]: Not necessary anymore, since `NtfsFindMftRecord` applies the mask (see PR #3775).
```suggestion
```

File: boot/freeldr/bootsect/ntfs.S
Comment (HBelusca) [inline boot/freeldr/bootsect/ntfs.S:413]: Those three lines seem to cause a loading problem (takes ~ 40 seconds); @DarkFire01 and @archeYR have a patch for it that we're currently reviewing.

File: boot/freeldr/bootsect/ntfs.S
Comment (HBelusca) [inline boot/freeldr/bootsect/ntfs.S:842]:
```suggestion
// Finds an attribute header into the MFT
```

File: boot/freeldr/bootsect/ntfs.S
Comment (HBelusca) [reply boot/freeldr/bootsect/ntfs.S:413]: @sdever Are you able to reproduce that problem I evoked above? Or does freeldr load fast for you?

File: boot/freeldr/bootsect/ntfs.S
Comment (disean) [inline boot/freeldr/bootsect/ntfs.S:350]: Can you document the location and size of the disk buffer? See https://git.reactos.org/?p=reactos.git;a=blob;f=boot/freeldr/bootsect/ext2.S;h=bd3119c635feafba68ee36677098b4aa3efdea0d;hb=refs/heads/master#l129

**Review by disean — COMMENTED** [review]
Minor suggestions for reducing code size

**Comment by HBelusca** [comment]
@sdever I think that we'll merge this PR soon, in order to not block it more (if you don't see anything against it).
Those extra TODOs can be dealt with in separate PRs.

File: boot/freeldr/bootsect/ntfs.S
Comment (DarkFire01) [reply boot/freeldr/bootsect/ntfs.S:416]: This has been resolved I think, I formatted a drive on XP and it worked as you’d expect. Unless I’m missing something can be resolved.

File: boot/freeldr/freeldr/lib/fs/ntfs.c
Comment (tkreuzer) [inline boot/freeldr/freeldr/lib/fs/ntfs.c:4]: Please revert this change.

**Review by tkreuzer — APPROVED** [review]

File: boot/freeldr/freeldr/lib/fs/ntfs.c
Comment (JoachimHenze) [reply boot/freeldr/freeldr/lib/fs/ntfs.c:4]: Timo, This is no intentional change by sdever. It is a bug in the github website, **which always chokes on Herves** name when showing PRs. That is why I recommended in the past to change his name to H. Poussineau to workaround that github bug. You can verify my claim by putting a .patch behind the URL
https://github.com/reactos/reactos/pull/3416.patch
The change is not visible in there.

Comment (tkreuzer) [reply boot/freeldr/freeldr/lib/fs/ntfs.c:4]: Huh? GitHub doesn't make up changes in unmodified files!
Also what is this?
`
  *  Copyright (C) 2004  Filip Navara  <xnavara@volny.cz>
- *  Copyright (C) 2009-2010  Herv� Poussineau
+ *  Copyright (C) 2009-2010  Herv� Poussineau
  *
`

Comment (JoachimHenze) [reply boot/freeldr/freeldr/lib/fs/ntfs.c:4]: You are right. It **is** within the patch. I have seen this exact same thing on many PRs before though.

Comment (binarymaster) [reply boot/freeldr/freeldr/lib/fs/ntfs.c:4]: It's not a problem and it doesn't need a change. File was encoded in Win-1252, and now it's UTF-8.

Comment (JoachimHenze) [reply boot/freeldr/freeldr/lib/fs/ntfs.c:4]: Thank you for double-checking this specific file @binarymaster .
We should double-check all others too with "Poussineau".

**Review by HBelusca — APPROVED** [review]

**Comment by sdever** [comment]
> @sdever I think that we'll merge this PR soon, in order to not block it more (if you don't see anything against it). Those extra TODOs can be dealt with in separate PRs.

Yes, it can be merged as is

**Review by HBelusca — APPROVED** [review]

**Review by binarymaster — APPROVED** [review]
🚢 :shipit:

**Review by DarkFire01 — APPROVED** [review]

**Comment by HBelusca** [comment]
Thanks for your work @sdever ! :tada:

**Pull request merged**
